src/signer_node/mod.rs

```diff
@@ -131,120 +131,145 @@ impl<T: TapyrusApi, C: ConnectionManager> SignerNode<T, C> {
-                        panic!("{:?}", e);
-                    }
-                },
-                None => {
-                    // Stop signal receiver is not set. Do nothing.
+            log::info!("Start thread for redis subscription");
+            let (sender, receiver): (Sender<Message>, Receiver<Message>) = channel();
+            let closure = move |message: Message| match sender.send(message) {
+                Ok(_) => ControlFlow::Continue,
+                Err(error) => {
+                    log::warn!("Happened error!: {:?}", error);
+                    ControlFlow::Break(())
+                }
+            };
+            let id = self.params.signer_id;
+            let handler = self.connection_manager.start(closure, id);
+
+            log::info!("Start Key generation Protocol");
+            // Idle 5s, before node starts Key Generation Protocol communication.
+            // To avoid that nodes which is late to startup can't receive messages.
+            log::info!("Idle 5 secs... ");
+            std::thread::sleep(Duration::from_secs(5));
+
+            // Start First Round
+            log::info!("Start block creation rounds.");
+            self.start_next_round();
+
+            loop {
+                match self.handle_signal() {
+                    Some(()) => return,
+                    None => {}
                 }
-            }
 
-            // Receiving message.
-            match receiver.try_recv() {
-                Ok(Message {
-                    message_type,
-                    sender_id,
-                    ..
-                }) => {
-                    log::debug!(
-                        "Got {} message from {:?}. MessageType: {:?}",
-                        message_type,
-                        sender_id,
-                        message_type
-                    );
-
-                    let next = self.process_round_message(&sender_id, message_type);
-                    self.current_state = next;
-
-                    if let NodeState::RoundComplete { .. } = &self.current_state {
-                        self.start_next_round()
-                    }
+                self.handle_message(&receiver);
 
-                    log::debug!("Current state updated as {:?}", self.current_state);
-                }
-                Err(TryRecvError::Empty) => {
-                    // No new messages. Do nothing.
+                self.handle_timer();
+
+                match self.handle_connection() {
+                    Some(_) => break,
+                    None => {}
                 }
-                Err(e) => log::debug!("{:?}", e),
+
+                // Wait for next loop 300 ms.
+                std::thread::sleep(Duration::from_millis(300));
+            }
+            log::info!("Wait for join thread {:?}", handler.thread().id());
+            match handler.join() {
+                Ok(_) => {}
+                Err(e) => log::warn!("Failed to join thread {:?}", e),
             }
```

Comment:
```suggestion
            if let Err(e) = handler.join() {
                log::warn!("Failed to join thread {:?}", e);
            }
```

Reply:
I think using `if let` is more simple.

Reply:
45d3efb4094b38bf2e2a76b1f7120db975497fe1